### PR TITLE
[kernel] New SpMM & SDDMM kernel on CPU and CUDA

### PR DESCRIPTION
**Pull request by yzh119**
## Description
First PR of DGL 0.5 kernel improvement (only C++ side).
CPU and CAPI part credits to @jermainewang .

## TODO
- [x] Simplifying broadcasting logic via utilizing stride information stored in dltensor.
- [ ] C++ side unit test.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- Deprecate minigun and write node/edge parallel by hand.
- Deprecate previous broadcasting logic (ravel & unravel function, which is costly).

### CODE REVIEW

**Comment by yzh119** [comment]
@jermainewang , @classicsong , have simplied the code and unified the bcast version and non-bcast version of SpMM and SDDMM. This PR is ready for review now.

File: src/kernel/cuda/sddmm.cuh [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+          lhsoff + lhs_add * reduce_size,
+          rhsoff + rhs_add * reduce_size,
+          reduce_size);
+      outoff[tx] = val;
+      tx += stride_x;
+    }
+    ty += stride_y;
+  }
+}
+
```
Comment (jermainewang) [inline src/kernel/cuda/sddmm.cuh:117]:
```
Macro used for switching between broadcasting and non-broadcasting kernels. It also copies
the auxiliary information for calculating broadcasting offsets to GPU.
```

File: src/kernel/cuda/sddmm.cuh [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    if ((EDGE_MAP)) {                                                          \
+      constexpr bool UseIdx = true;                                            \
+      { __VA_ARGS__ }                                                          \
+    } else {                                                                   \
+      constexpr bool UseIdx = false;                                           \
+      { __VA_ARGS__ }                                                          \
+    }                                                                          \
+  } else {                                                                     \
+    constexpr bool UseBcast = true;                                            \
+    DLContext ctx = (CTX);                                                     \
```
Comment (jermainewang) [inline src/kernel/cuda/sddmm.cuh:131]: `const DLContext& `

File: src/kernel/cuda/spmm.cuh [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+namespace cuda {
+
+template <typename T>
+__device__ __forceinline__ T _ldg(T* addr) {
+#if __CUDA_ARCH__ >= 350
+  return __ldg(addr);
+#else
+  return *addr;
+#endif
+}
```
Comment (jermainewang) [inline src/kernel/cuda/spmm.cuh:26]: This could be placed in `cuda_utils.h` in my PR.

File: src/kernel/cuda/spmm.cuh [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    const DType* eoff = BinaryOp::use_rhs ? (efeat + eid * efeat_len): nullptr;
+    DType* outoff = out + dst * out_len;
+    while (tx < out_len) {
+      const int64_t lhs_add = UseBcast ? ubcast_off[tx] : tx;
+      const int64_t rhs_add = UseBcast ? ebcast_off[tx] : tx;
+      DType val = BinaryOp::Call(uoff + lhs_add, eoff + rhs_add);
+      Idx* arguoff = (ReduceOp::require_arg && BinaryOp::use_lhs) ?
+        (arg_u + dst * out_len + tx): nullptr;
+      Idx* argeoff = (ReduceOp::require_arg && BinaryOp::use_rhs) ?
+        (arg_e + dst * out_len + tx): nullptr;
```
Comment (jermainewang) [inline src/kernel/cuda/spmm.cuh:56]: Is it true that for SpMMCoo, `arguoff` and `argeoff` are always nullptr?

File: src/kernel/cuda/spmm.cuh [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    const Idx src = _ldg(row + ty);
+    const Idx dst = _ldg(col + ty);
+    const Idx eid = UseIdx ? _ldg(edge_map + ty) : ty;
+    int64_t tx = blockIdx.x * blockDim.x + threadIdx.x;
+    const int64_t stride_x = blockDim.x * gridDim.x;
+    const DType* uoff = BinaryOp::use_lhs ? (ufeat + src * ufeat_len): nullptr;
+    const DType* eoff = BinaryOp::use_rhs ? (efeat + eid * efeat_len): nullptr;
+    const DType* outoff = out + dst * out_len;
+    Idx* arguoff = BinaryOp::use_lhs ? (arg_u + dst * out_len): nullptr;
+    Idx* argeoff = BinaryOp::use_rhs ? (arg_e + dst * out_len): nullptr;
```
Comment (jermainewang) [inline src/kernel/cuda/spmm.cuh:86]: And they will never be null?

File: src/kernel/cuda/spmm.cuh [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+      constexpr bool UseIdx = true;                                            \
+      { __VA_ARGS__ }                                                          \
+    } else {                                                                   \
+      constexpr bool UseIdx = false;                                           \
+      { __VA_ARGS__ }                                                          \
+    }                                                                          \
+    device->FreeWorkspace(ctx, (LHS_OFF));                                     \
+    device->FreeWorkspace(ctx, (RHS_OFF));                                     \
+  }                                                                            \
+} while (0)
```
Comment (jermainewang) [inline src/kernel/cuda/spmm.cuh:170]: Is this the same macro as the one in `sddmm.cuh`?

Comment (yzh119) [reply src/kernel/cuda/spmm.cuh:170]: Exactly

File: src/kernel/cuda/spmm.cuh (hunk 4 above)
Comment (yzh119) [reply src/kernel/cuda/spmm.cuh:56]: I don't understand why this is the case, shouldn't whether `argu/eoff` be depend on the selection of binary operator?

File: src/kernel/cuda/spmm.cuh (hunk 5 above)
Comment (yzh119) [reply src/kernel/cuda/spmm.cuh:86]: They will be null when we use `copy_u` or `copy_e` operator.

File: src/kernel/cuda/functor2.cuh [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    DType val, Idx uid, Idx eid) {
+    if (!atomic) {
+      *out_buf += val;
+    } else {
+      cuda::AtomicAdd(out_buf, val);
+    }
+  }
+  static __device__ __forceinline__ void CallArg(Idx fid,
+    Idx *arg_u_buf, Idx *arg_e_buf,
+    DType val, DType val_ref, Idx uid, Idx eid) {}
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/kernel/cuda/functor2.cuh:115]: Will get a compilation error if I remove this function.

File: src/kernel/cuda/spmm.cuh (hunk 6 above)
Comment (jermainewang) [reply src/kernel/cuda/spmm.cuh:170]: Then pls merge them

File: src/kernel/cuda/spmm.cuh (hunk 4 above)
Comment (jermainewang) [reply src/kernel/cuda/spmm.cuh:56]: But even for max and min reducers, you didn't use the argu and arge buffers right? Instead, you launch another kernel to compute them.

Comment (yzh119) [reply src/kernel/cuda/spmm.cuh]: You are right.

File: src/kernel/kernel2.cc [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    const std::vector<NDArray>& arrays,
+    const std::vector<std::string>& names) {
+  for (size_t i = 0; i < arrays.size(); ++i) {
+    if (aten::IsNullArray(arrays[i]))
+      continue;
+    CHECK_EQ(ctx, arrays[i]->ctx)
+      << "Expected device context " << ctx << ". But got "
+      << arrays[i]->ctx << " for " << names[i] << ".";
+  }
+}
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/kernel/kernel2.cc]: Not exactly the same case.

File: src/kernel/bcast.cc [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      for (int i = 1; i < std::max(dl, dr); ++i) {
+        for (int offset = 0; offset < out_len; ++offset) {
+          rst.lhs_offset.push_back(rst.lhs_offset[offset] + i * (i < dl) * stride_l);
+          rst.rhs_offset.push_back(rst.rhs_offset[offset] + i * (i < dr) * stride_r);
+        }
+      }
+      out_len *= std::max(dl, dr);
+      stride_l *= dl;
+      stride_r *= dr;
+    }
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/kernel/bcast.cc]: solved by enforcing the input tensors to have ndim>=2.
e.g. if ufeat has shape (num_nodes,), we reshape it to (num_nodes, 1).

File: src/kernel/bcast.cc [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    rst.out_len = out_len;
+  } else {
+    rst.out_len = (op == "copy_e") ? rst.rhs_len : rst.lhs_len;
+    if (op == "dot") {
+      rst.reduce_size = lhs->shape[lhs->ndim - 1];
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/kernel/bcast.cc]: ditto.

File: src/kernel/bcast.cc [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  rst.lhs_len = lhs->strides[0];
+  rst.rhs_len = rhs->strides[0];
+  rst.use_bcast = UseBcast(op, lhs, rhs);
+  rst.reduce_size = 1;
+  if (rst.use_bcast) {
+    const int max_ndim = std::max(lhs->ndim, rhs->ndim) - 1;
+    int out_len = 1, j = 0;
+    if (op == "dot") {
+      rst.reduce_size = lhs->shape[lhs->ndim - 1];
+      ++j;
```
[the remark this answers is not in the document]
Comment (yzh119) [reply src/kernel/bcast.cc]: ditto.

File: src/array/cpu/spmm.h [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+  const IdType* indptr = static_cast<IdType*>(csr.indptr->data);
+  const IdType* indices = static_cast<IdType*>(csr.indices->data);
+  const IdType* edges = has_idx ? static_cast<IdType*>(csr.data->data) : nullptr;
+  const DType* X = Op::use_lhs? static_cast<DType*>(ufeat->data) : nullptr;
+  const DType* W = Op::use_rhs? static_cast<DType*>(efeat->data) : nullptr;
+  const int64_t dim = bcast.out_len,
+                lhs_dim = bcast.lhs_len,
+                rhs_dim = bcast.rhs_len;
+  DType* O = static_cast<DType*>(out->data);
+  IdType* argX = Op::use_lhs? static_cast<IdType*>(argu->data) : nullptr;
```
Comment (classicsong) [inline src/array/cpu/spmm.h:140]: Is this for backward?
Can you add some doc about this?

File: src/array/cpu/spmm.h [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+  IdType* argW = Op::use_rhs? static_cast<IdType*>(arge->data) : nullptr;
+  const int64_t nnz = coo.row->shape[0];
+  // fill zero elements
+#pragma omp parallel for
+  for (IdType i = 0; i < out.NumElements(); ++i) {
+    O[i] = Cmp::zero;
```
Comment (classicsong) [inline src/array/cpu/spmm.h:211]: use memset?

File: src/array/cuda/spmm.cuh [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+ * \brief CUDA kernel to compute argu and arge in g-SpMM on Coo format.
+ * \note it uses edge parallel strategy, different threadblocks (on y-axis)
+ *       is responsible for the computation on different edges. Threadblocks
+ *       on the x-axis are responsible for the computation on different positions
+ *       in feature dimension.
+ */
+template <typename Idx, typename DType,
+          typename BinaryOp, typename ReduceOp,
+          bool UseBcast = false, bool UseIdx = false>
+__global__ void ArgSpMMCooKernel(
```
Comment (classicsong) [inline src/array/cuda/spmm.cuh:89]: what does arg stand for?

File: tests/compute/test_sparse.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    'copy_u': lambda edges: {'m': edges.src['x']},
+}
+
+udf_reduce = {
+    'sum': lambda nodes: {'v': F.sum(nodes.mailbox['m'], 1)},
+    'min': lambda nodes: {'v': F.min(nodes.mailbox['m'], 1)},
+    'max': lambda nodes: {'v': F.max(nodes.mailbox['m'], 1)}
+}
+
+def test_spmm():
```
Comment (classicsong) [inline tests/compute/test_sparse.py:29]: DId you add test for broadcast?

File: tests/compute/test_sparse.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    for msg in ['add', 'sub', 'mul', 'div', 'copy_u', 'copy_e']:
+        for reducer in ['sum', 'min', 'max']:
+            print('SpMM(message func: {}, reduce func: {})'.format(msg, reducer))
+            v = dgl.gspmm(g, msg, reducer, u, e)[0]
+            g.update_all(udf_msg[msg], udf_reduce[reducer])
+            v1 = g.ndata['v']
+            assert F.allclose(v, v1, rtol=1e-3, atol=1e-3)
+            print('passed')
+
+def test_sddmm():
```
Comment (classicsong) [inline tests/compute/test_sparse.py:48]: Same here.

File: tests/compute/test_sparse.py (hunk 15 above)
Comment (yzh119) [reply tests/compute/test_sparse.py:29]: We are testing on a broadcasting setting.

File: src/array/cpu/spmm.h (hunk 13 above)
Comment (yzh119) [reply src/array/cpu/spmm.h:211]: We don't know the data type of O, maybe std::fill is a better solution.
btw, I agree that it's not necessary to use openmp here.

File: src/array/cuda/spmm.cuh (hunk 14 above)
Comment (yzh119) [reply src/array/cuda/spmm.cuh:89]: It's used to compute arg_u and arg_e, which is useful in computing gradients of min/max reducer.

File: python/dgl/sparse.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    """
+    if u is not None:
+        if F.ndim(u) == 1:
+            u = F.unsqueeze(u, -1)
+    if e is not None:
+        if F.ndim(e) == 1:
+            e = F.unsqueeze(e, -1)
+
+    op = op_mapping[op]
+    ctx = F.context(u) if u is not None else F.context(e)
```
[the remark this answers is not in the document]
Comment (yzh119) [reply python/dgl/sparse.py:111]: We check the consistency of context on C++ end.

File: src/array/cpu/spmm.h (hunk 12 above)
Comment (yzh119) [reply src/array/cpu/spmm.h:141]: I have added some docstring describing arg_u and arg_e.

File: include/dgl/kernel.h [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+/*!
+ * \brief Generalized Sampled Dense-Dense Matrix Multiplication on Coo format.
+ */
+template <int XPU, typename IdType, typename DType>
+void SDDMMCoo(const std::string& op,
+              const BcastOff& bcast,
+              const aten::COOMatrix& coo,
+              NDArray ufeat,
+              NDArray efeat,
+              NDArray out);
```
Comment (jermainewang) [inline include/dgl/kernel.h:102]: ditto

File: include/dgl/runtime/ndarray.h [hunk 19]
```diff
@@ -130,7 +130,7 @@ class NDArray {
   /*! \return the data pointer with type. */
   template <typename T>
   inline T* Ptr() const {
-    if (!defined())
+    if (!defined() || operator->()->shape[0] == 0)
```
Comment (jermainewang) [inline include/dgl/runtime/ndarray.h:133]: I don't think that's a good idea. `Ptr` should just return the internal data pointer.

File: src/array/cpu/sddmm.h [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+/*!
+ * \brief CPU kernel of g-SDDMM on Csr format.
+ * \param bcast Broadcast information.
+ * \param csr The Csr matrix.
+ * \param ufeat The feature on source nodes.
+ * \param vfeat The feature on destination nodes.
+ * \param out The result feature on edges.
+ * \note it uses node parallel strategy, different threads are responsible
+ *       for the computation of different nodes. To avoid possible data hazard,
+ *       we use atomic operators in this case.
```
Comment (jermainewang) [inline src/array/cpu/sddmm.h:25]: I don't think you use atomic operators for SDDMM.

File: src/array/kernel.cc [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+inline void CheckShape(
+    const std::vector<uint64_t>& gdim,
+    const std::vector<int>& uev_idx,
+    const std::vector<NDArray>& arrays,
+    const std::vector<std::string>& names) {
+  for (size_t i = 0; i < arrays.size(); ++i) {
+    if (IsNullArray(arrays[i]))
+      continue;
+    CHECK_GE(arrays[i]->ndim, 2)
+      << "Expect " << names[i] << " to have ndim >= 2";
```
Comment (jermainewang) [inline src/array/kernel.cc:55]: "Expect ...; For scalar features, please expand its dimension with an additional dimension of length one."

File: tests/compute/test_sparse.py [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+        print('SDDMM(message func: {})'.format(msg))
+        e = dgl.gsddmm(g, msg, u, v)
+        g.apply_edges(udf_apply_edges[msg])
+        e1 = g.edata['m']
+        assert F.allclose(e, e1, rtol=1e-3, atol=1e-3)
+        print('passed')
+
+if __name__ == '__main__':
+    test_spmm()
+    test_sddmm()
```
Comment (jermainewang) [inline tests/compute/test_sparse.py:68]: Please test some corner cases:
* Features are of scalar shapes.
* Empty graph / graph with no edges.

Also miss tests on bipartite graphs.

File: include/dgl/runtime/ndarray.h (hunk 19 above)
Comment (yzh119) [reply include/dgl/runtime/ndarray.h:133]: ```operator->()->shape[0] == 0``` is the condition that an NDArray is NullArray.
In some kernel we need this behaviour (e.g. use `edge_map != nullptr` to decide whether edge mapping is required or not).

Comment (yzh119) [reply include/dgl/runtime/ndarray.h]: Removed all dependencies.
